File: android/core/domain/src/main/kotlin/com/akaar/core/domain/model/Product.kt
package com.akaar.core.domain.model

/**
 * The capability block. Every field here appears in the matching contract
 * (docs/02-interdependence/shared-spine.md section 4), which is the test for
 * whether a field belongs in it: a field no buyer filters on is a field the
 * seller should not be asked to fill.
 *
 * Complete before publication, or the product is unmatchable and therefore
 * invisible rather than merely incomplete.
 */
data class Capability(
    val moq: Int,
    val capacityPerCycle: Int,
    val cycleDays: Int,
    val leadTimeDays: Int,
    val madeToOrder: Boolean = false,
    val customizationSupported: Boolean = false,
    val customizationTypes: List<String> = emptyList(),
) {
    /**
     * The quantity this seller can actually deliver by [daysAvailable], mirroring
     * the SQL in search_marketplace. Kept here so the client can explain a
     * result rather than only display it.
     */
    fun deliverableBy(daysAvailable: Int): Int {
        val productionDays = daysAvailable - leadTimeDays
        if (productionDays < 0 || cycleDays <= 0) return 0
        val cycles = maxOf(1, productionDays / cycleDays)
        return capacityPerCycle * cycles
    }

    fun canMeet(quantity: Int, daysAvailable: Int): Boolean =
        quantity >= moq && quantity <= deliverableBy(daysAvailable)
}

/** Text carried in both languages. The source is never overwritten by a translation. */
data class Translatable(
    val hi: String? = null,
    val en: String? = null,
    val sourceLang: String = "hi",
) {
    fun inLang(lang: String): String? = when (lang) { "hi" -> hi ?: en; else -> en ?: hi }
    val source: String? get() = if (sourceLang == "hi") hi else en
}

/**
 * What a buyer can see. There is no cost, floor, margin or confidence field on
 * this type at all - those live in [SellerPricing], which is a separate table
 * with owner-only access, so the omission is structural rather than a matter of
 * remembering to strip it.
 */
data class Product(
    val id: String,
    val sellerId: String,
    val status: ProductStatus,
    val title: Translatable,
    val description: Translatable = Translatable(),
    val craftId: String?,
    val productType: String?,
    val materialIds: List<String> = emptyList(),
    val techniqueIds: List<String> = emptyList(),
    val colors: List<String> = emptyList(),
    val care: Translatable = Translatable(),
    val regionState: String?,
    val regionDistrict: String?,
    val capability: Capability?,
    val listedPrice: Money?,
    val negotiable: Boolean = true,
    val stockQty: Int = 0,
    val reservedQty: Int = 0,
    val b2bEnabled: Boolean = false,
    val media: List<Media> = emptyList(),
) {
    val availableQty: Int get() = (stockQty - reservedQty).coerceAtLeast(0)
    val originalImage: Media? get() = media.firstOrNull { it.role == MediaRole.Original }
    val displayImage: Media? get() = media.firstOrNull { it.role == MediaRole.Enhanced } ?: originalImage

    /**
     * The publication invariants, mirrored from publish_product() so the UI can
     * show what is still missing before the round trip. The server remains the
     * authority; this is an affordance, not a check.
     */
    fun missingForPublication(hasPriceCorridor: Boolean): List<String> = buildList {
        if (title.hi.isNullOrBlank() || title.en.isNullOrBlank()) add("title")
        if (craftId == null) add("craft")
        if (listedPrice == null) add("price")
        if (capability == null) add("capability")
        if (regionState == null || regionDistrict == null) add("region")
        if (originalImage == null) add("photo")
        if (!hasPriceCorridor) add("price_corridor")
    }
}

data class Media(
    val id: String,
    val role: MediaRole,
    val storageKey: String,
    val derivedFromId: String? = null,
    val sellerApproved: Boolean = false,
    /** Must be labelled wherever it is shown, in both the seller and buyer UI. */
    val isAiBackground: Boolean = false,
)

/**
 * Seller-only. The Fair Price Shield floor and the negotiation corridor.
 * Never fetched in a buyer session - the row is not readable by one.
 */
data class SellerPricing(
    val productId: String,
    val materialsCost: Money,
    val labourHours: Double,
    val labourRate: Money,
    val packagingCost: Money,
    val overheadCost: Money,
    val shippingEstimate: Money,
    val platformFeePct: Double,
    val minMarginPct: Double,
    val sustainableFloor: Money,
    val d2cRecommended: Money?,
    val wholesaleMin: Money?,
    val wholesaleMax: Money?,
    val netEarningsEstimate: Money?,
    val confidence: Double?,
    val explanation: Translatable = Translatable(),
    val preferredPrice: Money?,
    val minAcceptablePrice: Money?,
) {
    /**
     * The invariant that reconciles the two source documents: the private
     * minimum may never fall below the computed floor. Enforced as a CHECK
     * constraint in the database; repeated here so the UI can refuse earlier
     * and explain why.
     */
    fun isCorridorValid(): Boolean =
        minAcceptablePrice == null || minAcceptablePrice >= sustainableFloor

    fun clearsFloor(amount: Money): Boolean = amount >= sustainableFloor
}

File: mobile/packages/core-domain/src/product.ts
import type { Capability } from "./capability.ts";
import { MediaRole, type ProductStatus } from "./enums.ts";
import { Money } from "./money.ts";
import type { Translatable } from "./translatable.ts";

export interface Media {
  id: string;
  role: MediaRole;
  storageKey: string;
  derivedFromId?: string;
  sellerApproved?: boolean;
  /** Must be labelled wherever it is shown, in both the seller and buyer UI. */
  isAiBackground?: boolean;
}

/**
 * Port of the `Product` data class in android/core/domain/.../Product.kt.
 *
 * What a buyer can see. There is no cost, floor, margin or confidence field
 * on this type at all - those live in {@link SellerPricing}, which is a
 * separate table with owner-only access, so the omission is structural
 * rather than a matter of remembering to strip it.
 */
export interface Product {
  id: string;
  sellerId: string;
  status: ProductStatus;
  title: Translatable;
  description?: Translatable;
  craftId: string | null;
  productType: string | null;
  materialIds?: string[];
  techniqueIds?: string[];
  colors?: string[];
  care?: Translatable;
  regionState: string | null;
  regionDistrict: string | null;
  capability: Capability | null;
  listedPrice: Money | null;
  negotiable?: boolean;
  stockQty?: number;
  reservedQty?: number;
  b2bEnabled?: boolean;
  media?: Media[];
}

export function availableQty(product: Product): number {
  return Math.max(0, (product.stockQty ?? 0) - (product.reservedQty ?? 0));
}

export function originalImage(product: Product): Media | undefined {
  return (product.media ?? []).find((m) => m.role === MediaRole.Original);
}

export function displayImage(product: Product): Media | undefined {
  return (product.media ?? []).find((m) => m.role === MediaRole.Enhanced) ?? originalImage(product);
}

/**
 * The publication invariants, mirrored from `publish_product()` so the UI
 * can show what is still missing before the round trip. The server remains
 * the authority; this is an affordance, not a check.
 */
export function missingForPublication(product: Product, hasPriceCorridor: boolean): string[] {
  const missing: string[] = [];
  if (!product.title.hi?.trim() || !product.title.en?.trim()) missing.push("title");
  if (product.craftId == null) missing.push("craft");
  if (product.listedPrice == null) missing.push("price");
  if (product.capability == null) missing.push("capability");
  if (product.regionState == null || product.regionDistrict == null) missing.push("region");
  if (originalImage(product) == null) missing.push("photo");
  if (!hasPriceCorridor) missing.push("price_corridor");
  return missing;
}

/**
 * Seller-only. The Fair Price Shield floor and the negotiation corridor.
 * Never fetched in a buyer session - the row is not readable by one.
 */
export interface SellerPricing {
  productId: string;
  materialsCost: Money;
  labourHours: number;
  labourRate: Money;
  packagingCost: Money;
  overheadCost: Money;
  shippingEstimate: Money;
  platformFeePct: number;
  minMarginPct: number;
  sustainableFloor: Money;
  d2cRecommended: Money | null;
  wholesaleMin: Money | null;
  wholesaleMax: Money | null;
  netEarningsEstimate: Money | null;
  confidence: number | null;
  explanation?: Translatable;
  preferredPrice: Money | null;
  minAcceptablePrice: Money | null;
}

/**
 * The invariant that reconciles the two source documents: the private
 * minimum may never fall below the computed floor. Enforced as a CHECK
 * constraint in the database; repeated here so the UI can refuse earlier and
 * explain why.
 */
export function isCorridorValid(pricing: SellerPricing): boolean {
  return pricing.minAcceptablePrice == null || pricing.minAcceptablePrice.atLeast(pricing.sustainableFloor);
}

export function clearsFloor(pricing: SellerPricing, amount: Money): boolean {
  return amount.atLeast(pricing.sustainableFloor);
}
